### PR TITLE
deployment blog first draft

### PR DESCRIPTION
**Pull request by garyverhaegen-da**
I did not know where else to put this; I think it makes sense as part of this repo.

### CODE REVIEW

File: notes/deplpy-0.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
+# Designing DAML-Driven Systems, part 1: Where does DAML fit?
+
+This is the first in a series of blog posts about the design and deployment of
+DAML-Driven systems, i.e. focusing on how DAML interacts with the rest of your
```
Comment (bame-da) [inline notes/deplpy-0.md]: This first sentence is slightly ambiguous. maybe split it in two:
1. Intro to blog series -- this is what the series will cover
2. Intro to this post -- focussing on DAML's place within the app

File: notes/deplpy-0.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+- By specifying a path to a DAR file as a (repeatable) command-line option to
+  the sandbox directly, or
+- After startup, by uploading a DAR file through the gRPC API.
+
+There is no plan for another ephemeral storage at this point.
+
+### Centralized storage
+
+While the DAML Language has been designed to make it easy to express
+multi-party, distributed workflows, it consequently also makes it really easy
```
Comment (cocreature) [inline notes/deplpy-0.md]: Why does that follow? If you optimize for solving the distributed case, the non-distributed case might be harder than something that ignores the distributed case.

File: notes/deplpy-0.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+to describe non-distributed workflows. We believe the DAML way of modeling data
+can be beneficial for most applications and therefore there is a place for DAML
+models even in single-stakeholder scenarios.
+
+To cater to those use-cases, we currently offer a [PostgreSQL-backed persistent
+version of the sandbox](), also as part of the open-source [DAML SDK]().
+
+Note that in this context "centralized" is meant in the DAML meaning, i.e.
+there is only one party that unilaterally controls the data store. It is
```
Comment (cocreature) [inline notes/deplpy-0.md]: This is a bit confusing, you talk about “DAML meaning” and then use “party” in a way that does not correspond to a DAML party afaict.

File: notes/deplpy-0.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+The best technology to deliver such a shared database is known as a distributed
+ledger. There are many mature, stable implementations of distributed ledgers
+with different tradeoffs and design goals (e.g. throughput vs. latency or
+different choices of supported programming languages); DAML provides a
+higher-level, portable way of describing workflows to run on many of them.
+
+If you're considering a distributed use-case, please refer to the [DAML-on-X]()
+documentation for an up-to-date list of currently-supported DAML backends, as
+we and our partners are constantly working hard at adding more.
```
Comment (cocreature) [inline notes/deplpy-0.md]:
```suggestion
we and our partners are constantly working hard on adding more.
```

File: notes/deplpy-0.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+the DAML Runtime itself as well as enabling high-performance interactions with
+it.
+
+However, this API has two main drawbacks:
+
+- The gRPC transport is not natively supported by many languages and platforms;
+  in many cases this means adding extra native dependencies to user code, which
+  can add complexity to build and deployment pipelines, especially for managed
+  code (e.g. Python, JS, Ruby, Java, etc.), but it also outright prevents some
+  use-cases such as direct browser access.
```
Comment (cocreature) [inline notes/deplpy-0.md]: I’m not sure if that is true, see https://github.com/grpc/grpc-web

File: notes/deplpy-0.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
+# Designing DAML-Driven Systems, part 1: Where does DAML fit?
```
[the remark this answers is not in the document]
Comment (garyverhaegen-da) [reply notes/deplpy-0.md]: I noticed that after posting the links on Slack and didn't want to break them ;-)

I will correct it.

File: notes/deplpy-0.md (hunk 5 above)
Comment (garyverhaegen-da) [reply notes/deplpy-0.md]: Maybe the phrasing is a bit too strong, but I think it is factually true: grpc-web relies on you placing an HTTP-to-gRPC translation proxy in-between your server(s) and the browser, so it's no longer a direct connection.

I'll rephrase it to be a bit less strong.

File: notes/deplpy-0.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+## What is DAML?
+
+To understand where DAML fits, it is useful to start by taking a step back and
+taking a look at what DAML is, from an architectural perspective. DAML is
+really two things:
+
+1. *The DAML Language*: A domain-specific language tailored to describe
+   multi-party workflows, i.e. permissioned state machines.
+2. *The DAML Runtime*: A runtime engine that can take such a set of DAML
```
[the remark this answers is not in the document]
Comment (garyverhaegen-da) [reply notes/deplpy-0.md]: It was not my intention to suggest any specific shape yet at this point; as a point of reference, "the language" doesn't really exist as a separate process at runtime either.

Regarding the `DAML engine` confusion, it is indeed unfortunate. Happy to take any suggestion for a better name for "a system that exposes a DAML API".

File: notes/deplpy-0.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+In the context of this series, we will not look further at the DAML language
+itself, and instead focus on the DAML Runtime. We will call "frontend" any
+application or system that talks to the generated API (they can range from
+automated systems to graphical interfaces for direct user input) and "backend"
+any compatible storage system.
+
+## Authentication
+
+The DAML Language has powerful, first-class support for dynamically defining
+user roles on a per-state-machine-instance basis, but has no concept of
```
[the remark this answers is not in the document]
Comment (garyverhaegen-da) [reply notes/deplpy-0.md]: I did not think that was needed; I will add a bit more text to link my verbiage to DAML concepts (as much as I understand them).

File: notes/deplpy-0.md (hunk 2 above)
Comment (garyverhaegen-da) [reply notes/deplpy-0.md]: I agree it does not automatically follow in the general case. I do believe it is true in this case, though; at least in my experience DAML is the best language I know to define workflows, even in the single-user case. That said, I have not looked at many others in a while. I'll update the wording.

File: notes/deplpy-0.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply notes/deplpy-0.md]: I don’t know myself what exactly this refers to so I think it deserves at least a link :slightly_smiling_face:

File: notes/deplpy-0.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply notes/deplpy-0.md]: I’ll see if I can come up with better wording.

File: notes/deplpy-0.md (hunk 5 above)
Comment (cocreature) [reply notes/deplpy-0.md]: Oh I didn’t realize this needs a proxy, yeah in that case this is totally fine in a slightly less strong wording.

File: notes/deplpy-0.md (hunk 2 above)
Comment (cocreature) [reply notes/deplpy-0.md]: I do agree with this for DAML as well, I just found the language a bit too general.

File: notes/deplpy-0.md (hunk 3 above)
Comment (garyverhaegen-da) [reply notes/deplpy-0.md]: I'm referring to an approach @bame-da suggested, where, in the case where there is one operator with total control of the underlying storage, we reify that at the DAML model level by making it the single signatory on all instances, thereby making data updates trivial.

Does that added context make this paragraph make sense? If not, I need to revisit my understanding of what a DAML party is. If so, should I add that context in the post?

Comment (cocreature) [reply notes/deplpy-0.md]: I think it would be fine to suggest that it is possible to reify the topology in DAML or even recommended but this paragraph suggests to me that this happens automatically and always if you run sandbox on postgres. If you consider things like DAVL which runs on postgres, we do not follow this approach ourselves and overall I would say very few of our applications are written this way. It also goes against the idea of being able to write a DAML app that you initially run on postgres and then later switch to a distributed ledger if the topology of your ledger is baked into your model.

In addition to that, there are also interesting deployments like DABL where if you want to have the centralized aspect represented in your model, the operator is probably the application developer (e.g. @bame-da for Danban) and not the people running DABL.

Comment (garyverhaegen-da) [reply notes/deplpy-0.md]: I've refrained from touching this for now — I think there should be a separate blog about upgrading considerations, and this should be discussed there.

**Comment by garyverhaegen-da** [comment]
This is almost two years old, I don't think it's worth keeping it open at this point.
